third_party/txt/src/txt/paragraph_txt.cc

```diff
@@ -354,11 +355,40 @@ bool ParagraphTxt::ComputeLineBreaks() {
 
     size_t breaks_count = breaker_.computeBreaks();
     const int* breaks = breaker_.getBreaks();
-    for (size_t i = 0; i < breaks_count; ++i) {
-      size_t break_start = (i > 0) ? breaks[i - 1] : 0;
-      size_t line_start = break_start + block_start;
-      size_t line_end = breaks[i] + block_start;
-      bool hard_break = i == breaks_count - 1;
+    for (size_t i = 0; i < breaks_count && !early_ellipsized; ++i) {
+      size_t break_start;
+      size_t line_start;
+      size_t line_end;
+      bool hard_break;
+      double line_width = breaker_.getWidths()[i];
+      // We only want to early ellipsized if user choose ellipsizing by character.
+      // Other type of ellipsizing should still parse through the paragraph.
+      if (paragraph_style_.ellipsized_by_character &&
```

Comment:
Would it be simpler to do this as a post processing step?

After adding the `line_ranges_` for this block:
  * (if unlimited lines) check if `breaks_count > 1`, indicating that the first line is a soft break
  * (if `max_lines` is set) check whether the line range at index `max_lines - 1` is a soft break

If so, then merge the soft break line with the following line and flag it as being ellipsized.